src/notebook/widget.ts

```diff
@@ -384,6 +395,48 @@ class NotebookWidget extends Widget {
   }
 
   /**
+   * Handle DOM events for the widget.
+   */
+  handleEvent(event: Event) {
+    switch (event.type) {
+    case 'blur':
+      this.model.mode = 'command';
+      break;
+    case 'focus':
+      this.model.mode = 'edit';
+      // Activate the appropriate cell.
+      let layout = this._notebook.layout as PanelLayout;
+      for (let i = 0; i < layout.childCount(); i++) {
+        let widget = layout.childAt(i) as BaseCellWidget;
+        if (widget.node.contains(event.target as HTMLElement)) {
+          this.model.activeCellIndex = i;
+          break;
+        }
+      }
+      break;
+    }
+  }
+
+  /**
+   * Initialize a cell widget.
+   */
+  protected initCell(cell: BaseCellWidget): void {
+    cell.input.editor.node.addEventListener('focus', this, true);
```

Comment:
Since focus events bubble, we should just add a single event listener to the notebook widget, instead of one to each cell widget.